staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -404,6 +406,19 @@ func isDeletionDup(a, b *Delta) *Delta {
 	return b
 }
 
+func isSyncDup(a, b *Delta) *Delta {
```

Comment:
hm, ok yeah I can imagine something here being very useful, but--

a) you have to consider the case where the name stayed the same but the UID changed and not suppress either event
b) you have to consider that there could be arbitrary disconnections/reconnections -- I'm not convinced that the case in the if statement is actually impossible

Reply:
Thanks @lavalamp @yliaog  appreciate for your reviews.

I found there is a problem of this PR that we can not distinguish this Sync Delta comes from `resync` or `reslist`. 

In the `resync` process, I believe this Sync Delta could be deduplicated, but in the `relist` process, we should hold this Sync Delta. Unfortunately, we can not distinguish it.

Thanks again. I will close this PR & issue.

Reply:
I think it might still be possible to dedup if the sync is e.g. exactly the same as an add or update!

It's fine to dedup something from a resync -- the fact that two items are in the queue means that it won't go unprocessed. Might need to survey existing usage to verify this though...

Reply:
yes, it's a good idea to dedup adjacenct deltas own exactly same obj.

There is one situations in my mind while multi dis/re connections happening continuously, which leads to e.g. two same Sync delta are adjacent for exactly one Object(ObjA), like `<Sync,ObjA>、<Sync.ObjA>` in DeltaFIFO(items).

Are there any situations that adjacenct deltas own exactly same obj? :)

Besides, I think the adjacenct deltas must own different Obj if there are only watch & resync processes (after initial list). Is there any mistake in my understanding？

Reply:
Deltas are effectively sorted by object before we get here.

One change that should be clearly safe is to dedup two sync's with the exact same RV.

Reply:
Thanks @lavalamp  for your quick reply.

We care about two cases to conduct deduplication:
## 1. the adjacent Sync/Replaced deltas are exactly identical 
Specifically, both Delta.Type & Delta.Object are same, and in this case, we keep the newer one (ps: both are available). This case could happen after arbitrary disconnections/reconnections. 

## 2. Sync Delta only from Resync process
while the gate `emitDeltaTypeReplaced` is opened, it's clearly that Sync Delta definitely comes from Resync(), since Replace() can only emit Replaced Delta.  Then the following cases will dedup Deltas:
- case1 : `<Sync, ObjA>, <Deleted, ObjA'>`  ---- dedup to ----> ` <Deleted, ObjA'>`
- case2 : `<Sync, ObjA>, <Updated, ObjA'> ` ---- dedup to ---->  `<Updated, ObjA'>`
- case3 : `<Sync, ObjA>, <Added, ObjA'> ` ---- dedup to ----> `<Added, ObjA'>`
- case4 : `<Sync, ObjA>, <Replaced, ObjA'>` ---- dedup to ---->  `<Replaced, ObjA'>`
- case5 : `<Sync, ObjA>, <Replaced, ObjA> ` ---- dedup to ---->  `<Replaced, ObjA>`
> NOTE: ObjA' is newer than ObjA, which means ResourceVersion of ObjA' is larger than ObjA
		
The Reasons why it could conduct deduplication in above cases are: 
1) the purpose of Resync is helping listener to reconcile the whole object in knownObjects periodically, thus the most important thing is guaranteeing the whole objects appear in DeltaFIFO periodically, no matter how many Deltas for one object in DeltaFIFO.items. 
2) it should guarantee that `Delta.Object`s with new ResourceVersion should not be deduplicated, in other words, all the new `Delta.Object`s that not observed by the listeners should be processed. In the above cases, the object of Sync Delta <Sync, ObjA> must be already processed before, since the ObjA comes from knownObjects(Indexer).